### PR TITLE
Stricter test for glms with newton-cholesky solver on collinear data

### PR DESCRIPTION
**Pull request by ogrisel**
This PR aims at better testing the `lbfgs_step` fallback mechanism of #23314 (introduced in c9b120063574bdd9d408805ce96c34ebe7722f81 and subsequent commits).

This test is currently failing and highlights the fact that newton-cholesky with an inner solver that switches to 4 steps of LBFGS whenever the Hessian is found the ill-conditioned does not work as expected: it can be very slow because this problem can happen many times in a fit call and furthermore repeatedly calling 'L-BFGS-B' with `"maxiter": 4` via `scipy.optimize.minimize` is not equivalent to calling it once with a large `maxiter` because each time we loose the memory of the previous gradients.

I would therefore recommend to stop attempting to implement a fined grained fallback mechanism in the for the `inner_solve` step of the Newton solvers. Instead I would suggest to let the Newton solver raise an exception when this happens and use a coarse grained fallback to the "lbfgs" solver until convergence (possibly warm starting from the previous coef in case the newton solver had a chance to update them successfully for a few iterations). This coarse fallback should be much simpler to implement and maintain and should be guaranteed to converge to solution as good as the ones if the user had chosen LBFGS originally.

The scikit-learn-level warning message should be adapted accordingly.

What do you think about this plan @lorentzenchr?

Credits: this plan was originally suggested by @gaelvaroquaux IRL.

### CODE REVIEW

File: sklearn/linear_model/_glm/tests/test_glm.py [hunk 1]
```diff
@@ -1059,28 +1059,65 @@ def test_linalg_warning_with_newton_solver(global_random_seed):
...
+    # informative warning and fallback to the LBFGS solver
     msg = (
         "The inner solver of CholeskyNewtonSolver stumbled upon a"
         " singular or very ill-conditioned hessian matrix"
     )
     with pytest.warns(scipy.linalg.LinAlgWarning, match=msg):
-        PoissonRegressor(solver="newton-cholesky", alpha=0.0).fit(X_collinear, y)
+        reg = PoissonRegressor(solver="newton-cholesky", alpha=0.0, tol=tol).fit(
+            X_collinear, y
+        )
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/tests/test_glm.py:1008]: Note: I have checked that `LinAlgWarning` is also raised when replacing `solver="newton-cholesky"` by `solver="newton-qr-cholesky"` in this line.

We could probably parametrize the test with a `newton_solver` variable that takes values in `["newton-cholesky", "newton-qr-cholesky"]`.

**Comment by ogrisel** [comment]
I you agree with this plan, let me know and I can help you implement it to get the test in this PR to pass. But feel free to do it yourself if you prefer.

File: sklearn/linear_model/_glm/tests/test_glm.py [hunk 2]
```diff
@@ -1059,28 +1059,65 @@ def test_linalg_warning_with_newton_solver(global_random_seed):
...
     # No warning raised on well-conditioned design, even without regularization.
+    tol = 1e-10
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
+        reg = PoissonRegressor(solver="newton-cholesky", alpha=0.0, tol=tol).fit(
+            X_orig, y
+        )
+    original_newton_deviance = mean_poisson_deviance(y, reg.predict(X_orig))
+
+    # We check that the model could successfully overfit information in X_orig
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/tests/test_glm.py]:
```suggestion
    # We check that the model could successfully fit information in X_orig
```

File: sklearn/linear_model/_glm/tests/test_glm.py [hunk 3]
```diff
@@ -1059,28 +1059,65 @@ def test_linalg_warning_with_newton_solver(global_random_seed):
...
-    reg = PoissonRegressor(solver="newton-cholesky", alpha=1e-12).fit(X_collinear, y)
+    with warnings.catch_warnings():
+        warnings.simplefilter("error", scipy.linalg.LinAlgWarning)
+        PoissonRegressor(solver="newton-cholesky", alpha=1e-10).fit(X_collinear, y)
 
-    # Since we use a small penalty, the deviance of the predictions should still
-    # be almost the same.
-    this_deviance = mean_poisson_deviance(y, reg.predict(X_collinear))
-    assert this_deviance == pytest.approx(reference_deviance)
+    # While for most random seed the deviance of this model is very close to
```
Comment (ogrisel) [inline sklearn/linear_model/_glm/tests/test_glm.py]:
```suggestion
    # While for most random seeds the deviance of this model is very close to
```

**Comment by lorentzenchr** [comment]
> I would therefore recommend to stop attempting to implement a fined grained fallback mechanism in the for the `inner_solve` step of the Newton solvers. [...] What do you think about this plan?

**I'm very happy with it.** I thought a lot about the best action (in a user's perspective) of a solver in case of "dynamically detected" convergence problems:
1. Stop and raise error
2. **Warn and try different solver**, e.g. lbfgs (your suggestion)
3. Warn and temporarily use a different step in this iteration
  a. gradient step
  b. lbfgs step(s)
  c. step with modified hessian (e.g. adding a multiple of the identity=penalty, LDL decomposition, QR, SVD, ...)
...

**Comment by lorentzenchr** [comment]
> I you agree with this plan, let me know and I can help you implement it

I would appreciate your help very much. I added you as collaborator to this fork to make it easier to work together.

**Comment by ogrisel** [comment]
I will try to move this forward this afternoon. My plan to:

- first make the existing test pass on #23314 after the merge with `main`;
- merge this PR with an xfail and a TODO;
- implement the estimator-level fallback to LBFGS directly in #23314.

EDIT: @lorentzenchr was too fast. Let me update this PR.

**Comment by ogrisel** [comment]
I merged and now the updated collinear data test passes!
